cmd/admin-handlers.go

```diff
@@ -1303,6 +1304,154 @@ func (a adminAPIHandlers) KMSKeyStatusHandler(w http.ResponseWriter, r *http.Req
 	writeSuccessResponseJSON(w, resp)
 }
 
+// KMSKeyRotateHandler - POST /minio/admin/v2/kms/key/rotate
+func (a adminAPIHandlers) KMSKeyRotateHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := newContext(r, w, "KMSKeyRotateHandler")
+
+	objectAPI, _ := validateAdminReq(ctx, w, r, iampolicy.KMSKeyRotateAdminAction)
+	if objectAPI == nil {
+		return
+	}
+
+	if GlobalKMS == nil {
+		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrKMSNotConfigured), r.URL)
+		return
+	}
+
+	var req madmin.KMSRotateKeyRequest
+	if err := json.NewDecoder(io.LimitReader(r.Body, 1<<20)).Decode(&req); err != nil {
+		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+		return
+	}
+
+	// Delete rotation entries where the old key ID
+	// is equal to the new key ID. It's pointless to
+	// rotate when the old and the new key are the same.
+	for k, v := range req.KeyMapping {
+		if k == v {
+			delete(req.KeyMapping, k)
+		}
+	}
+
+	// Verify that the key mapping is well-formed. In particular,
+	// that a old key ID does not appear as new key ID as well.
+	// This is not allowed since it would cause unpredictable
+	// behavior.
+	for oldKey := range req.KeyMapping {
+		for _, newKey := range req.KeyMapping {
+			if oldKey == newKey {
+				err := AdminError{
+					StatusCode: http.StatusBadRequest,
+					Code:       http.StatusText(http.StatusBadRequest),
+					Message:    fmt.Sprintf("invalid key mapping: cannot rotate from and to: %s", oldKey),
+				}
+				writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+				return
+			}
+		}
+	}
+
+	// If there is no key mapping for rotation return early.
+	if len(req.KeyMapping) == 0 {
+		writeSuccessResponseHeadersOnly(w)
+		return
+	}
+
+	// If the bucket is empty and rotation is not recursive there are no
+	// objects - only buckets. So return early.
+	if req.Bucket == "" && !req.Recursive {
+		writeSuccessResponseHeadersOnly(w)
+		return
+	}
+
+	// Either check whether the requested bucket exists or
+	// if no bucket is specified list all buckets.
+	var bucketNames []string
+	if req.Bucket != "" {
+		if _, err := objectAPI.GetBucketInfo(ctx, req.Bucket); err != nil {
+			writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+			return
+		}
+		bucketNames = []string{req.Bucket}
+	} else {
+		bucketInfo, err := objectAPI.ListBuckets(ctx)
+		if err != nil {
+			writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
+			return
+		}
+		bucketNames = make([]string, len(bucketInfo))
+		for i, b := range bucketInfo {
+			bucketNames[i] = b.Name
+		}
+	}
+
+	// For each bucket list all objects for the requested prefix.
+	// If the object is encrypted with the KMS perform a key
+	// rotation if there is a mapping for the current key ID.
+	for _, bucket := range bucketNames {
```

Comment:
This seems quite fragile to me:

1) The client will time out in most cases. Expecting all objects in a bucket to be rotated within a request isn't feasible IMO.

2) What happens if this operation get re-tried? Is it idempotent?

3) What happens if the operation is interrupted? (crash)  Is there a way to resume?

Looking at it from an API perspective this should IMO be async and controlled more closely by the client.

Reply:
> The client will time out in most cases. Expecting all objects in a bucket to be rotated within a request isn't feasible IMO.

Hu?! Why? Note that a key rotation does not affect the object content. Also, why can't that be a long-running job? Why would the client timeout? It gets a response stream as long as the operation is running.

> What happens if this operation get re-tried? Is it idempotent?

No, it's not idempotent. But the point of a key rotation is to "rotate" e.g. all objects encrypted with `master-key-1` to `master-key-2`. If you run the operation again it will do "nothing" since no object is encrypted with `master-key-1`

> What happens if the operation is interrupted? (crash) Is there a way to resume?

Only a subset of objects will be rotated. If you resume you would rotate the remaining objects - which is AFAICS the "expected behavior".

> Looking at it from an API perspective this should IMO be async and controlled more closely by the client.

What do you mean by "more closely"?